fairseq/training.py

```diff
@@ -0,0 +1,490 @@
+        disallowed_in_config = ["save_interval", "max_epoch"]
+
+        for group in self.config_dict:
+            if isinstance(self.config_dict[group], dict):
+                for key in disallowed_in_config:
+                    assert (
+                        self.config_dict[group].get(key) is None
+                    ), f"please define {key} of {group} only as parameter in the post_config_dict"
+            else:
+                assert (
+                    self.config_dict[group].get(group) is None
+                ), f"please define {group} only as parameter in the post_config_dict"
+
+    def _sis_hash(self):
+        h = {"fairseq_hydra_config": self.config_dict}
+        return sis_hash_helper(h)
+
+
+class PytorchHydraModel:
+    """
+    Defines a Pytorch hydra model as yaml config, pytorch checkpoint file and epoch
+    """
+
+    def __init__(self, fairseq_hydra_config_file: tk.Path, model: tk.Path, epoch: int):
+        """
+        :param fairseq_hydra_config_file: Path to a returnn config file
+        :param model: Path to a pytorch checkpoint
+        :param epoch: Number of epochs this model was trained
+        """
+        self.returnn_config_file = fairseq_hydra_config_file
+        self.model = model
+        self.epoch = epoch
+
+
+class FairseqHydraTrainingJob(Job):
+    """
+    Train a Fairseq model using fairseq-hydra-train
+    """
+
+    def __init__(
+        self,
+        fairseq_hydra_config,
+        *,  # args below are keyword only
+        command_line_args=None,
+        max_epoch=1,
+        save_interval=1,
+        keep_epochs=None,
+        time_rqmt=4,
+        mem_rqmt=4,
+        cpu_rqmt=2,
+        gpu_rqmt=1,
+        fairseq_python_exe=None,
+        fairseq_root,
+        use_cache_manager=True,
+        zipped_audio_dir=None,
+    ):
+        """
+        :param FairseqHydraConfig fairseq_hydra_config:
+        :param list command_line_args: Additional command line arguments (starting with "--*"),
+            to configure the Fairseq-hydra task
+        :param int max_epoch: maximum number of epochs to run.
+        :param int save_interval: save a checkpoint each n-th epoch
+        :param list[int]|set[int]|None keep_epochs: specify which checkpoints are kept in self.out_models.
+            Use None for each save_interval-th epoch
+        :param int|float time_rqmt: Overall time requirements
+        :param int|float mem_rqmt: Memory requirements (per GPU)
+        :param int cpu_rqmt: Required number of CPUs (per GPU)
+        :param int gpu_rqmt: Number of required GPUs
+        :param tk.Path fairseq_python_exe: File path to the executable for running python
+        :param tk.Path fairseq_root: File path to the fairseq git for alternative call of fairseq-hydra-train
+            (no need to install fairseq here)
+        :param bool use_cache_manager: enables caching of data given in the manifest with the i6 cache manager
```

Comment:
We added it because other jobs that do caching use either RASR or RETURNN, both of which support caching. With fairseq, this is different. I'm not 100% sure about the details, but most likely it will directly read the files from python and certainly it does not know about the i6 cache manager. That's why we cached the files at the beginning of the `run` task and use an updated manifest with the cached paths when starting the fairseq training.

Reply:
Couldn't this be fixed by replacing the explicit calls to cf with `gs.file_caching` (and rename this arg to something like use_file_caching). Then whatever caching implementation the user specified in it's settings.py is used.

Reply:
I'm not very familiar with `gs.file_caching`. In the case here, we have a text file with all audio paths inside that will be read by fairseq. I'm not sure if this would be handled by sisyphus settings.

Reply:
The function `gs.file_caching` takes as input a `str` (the original location) and returns a `str` (the cached location). In between the function does the caching. [Default implementation](https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/global_settings.py#L107-L117) is to just return the original string.
It is invoked whenever one calls [`tk.Path.get_cached_path`](https://github.com/rwth-i6/sisyphus/blob/e07c227fbcbcb2e6e5b6ad03ce5ff87c30a90026/sisyphus/job_path.py#L188).

I think one could replace all calls to ```sp.check_output(["cf", something]).strip().decode("utf8")``` with ```gs.file_caching(something)``` except maybe the bundle file, which is a specialty of the i6 cache manager.


To clarify my point: I to not insist on changing it to be i6 independent, I merely start the discussion if we maybe think this is more elegant..

Reply:
Clearly this is not a very elegant solution. However, fairseq reads the audio files from a txt file. So we cache the audios and modify the txt file to contain the cached paths. It is not clear to me how sisyphus could interfere in this process of fairseq reading the filenames from the txt file. Or how we could modify the contents of the txt file such that fairseq would automatically use caching while reading.

Reply:
I don't have a much more elegant solution. However the solution I propose (see other comment below in run function) would be more generic as it would allow other places to insert their own version of caching.